framework/src/state_machine/event_queue.ts

```diff
@@ -59,18 +59,19 @@ export class EventQueue {
 		});
 	}
 
-	public createSnapshot(): void {
+	public createSnapshot(): number {
 		this._snapshotIndex = this._events.length;
+		return this._snapshotIndex;
 	}
 
-	public restoreSnapshot(): void {
-		const newEvents = this._events.splice(this._snapshotIndex);
+	public restoreSnapshot(snapshotID: number): void {
+		const newEvents = this._events.splice(snapshotID);
```

Comment:
this way of generating snapshotID is quite smart. more efficient and straight forward than managing in map 🚀

Reply:
Thanks!!